core/imagefs_linux.go

```diff
@@ -0,0 +1,82 @@
+//go:build linux
+// +build linux
+
+/*
+Copyright 2021 Mirantis
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package core
+
+import (
+	"syscall"
+	"time"
+
+	"github.com/docker/docker/api/types"
+	"github.com/sirupsen/logrus"
+
+	runtimeapi "k8s.io/cri-api/pkg/apis/runtime/v1"
+)
+
+// ImageFsInfo returns information of the filesystem of docker data root.
+func (ds *dockerService) imageFsInfo() (*runtimeapi.ImageFsInfoResponse, error) {
+	// collect info of the filesystem on which docker root resides
+	stat := &syscall.Statfs_t{}
+	err := syscall.Statfs(ds.dockerRootDir, stat)
+	if err != nil {
+		logrus.Error(err, "Failed to get filesystem info for %s", ds.dockerRootDir)
+		return nil, err
+	}
+	usedBytes := (stat.Blocks - stat.Bfree) * uint64(stat.Bsize)
+	iNodesUsed := stat.Files - stat.Ffree
+	logrus.Infof("Filesystem usage containing '%s': usedBytes=%v, iNodesUsed=%v", ds.dockerRootDir, usedBytes, iNodesUsed)
+
+	// compute total used bytes by docker images
+	images, err := ds.client.ListImages(types.ImageListOptions{All: true, SharedSize: true})
+	if err != nil {
+		logrus.Error(err, "Failed to get image list from docker")
+		return nil, err
+	}
+	var totalImageSize uint64
+	sharedSizeMap := make(map[int64]struct{})
+	for _, i := range images {
+		if i.SharedSize == -1 {
+			totalImageSize += uint64(i.Size)
+		} else { // docker version >= 23.0
+			totalImageSize += (uint64(i.Size) - uint64(i.SharedSize))
+			sharedSizeMap[i.SharedSize] = struct{}{}
```

Comment:
Yes, you are right. However, this is all I currently can do because I could not find the image layer chain info in the data returned from `docker images`.  This way is still better than what cadvisor stats does: cadvisor does not de-dup the shared size (at least in kube 1.24.6).

Reply:
Sounds good. I just wanted to understand our current limitations.